Mathlib/GroupTheory/Archimedean.lean

```diff
@@ -90,8 +156,113 @@ theorem AddSubgroup.cyclic_of_isolated_zero {H : AddSubgroup G} {a : G} (h₀ :
   · exact ⟨0, closure_singleton_zero.symm⟩
   · exact (exists_isLeast_pos hbot h₀ hd).imp fun _ => cyclic_of_min
 
+lemma AddSubgroup.isLeast_of_closure_iff_eq_max {a b : G} :
+    IsLeast {y : G | y ∈ closure ({a} : Set G) ∧ 0 < y} b ↔ b = max a (-a) ∧ 0 < b := by
+  constructor <;> intro h
+  · have := cyclic_of_min h
+    have ha : a ∈ closure ({b} : Set G) := by
+      simp [← this]
+    rw [mem_closure_singleton] at ha
+    obtain ⟨n, rfl⟩ := ha
+    have := h.left
+    simp only [mem_closure_singleton, mem_setOf_eq, ← mul_zsmul] at this
+    obtain ⟨m, hm⟩ := this.left
+    have key : m * n = 1 := by
+      rw [← (zsmul_strictMono_left this.right).injective.eq_iff, hm, one_zsmul]
+    rw [Int.mul_eq_one_iff_eq_one_or_neg_one] at key
+    rw [eq_comm]
+    rcases key with ⟨rfl, rfl⟩|⟨rfl, rfl⟩ <;>
+    simp [this.right.le, this.right]
+  · wlog ha : 0 ≤ a generalizing a
+    · convert @this (-a) ?_ (by simpa using le_of_not_le ha) using 4
+      · simp
+      · rwa [max_comm, neg_neg]
+    rw [max_eq_left ((neg_nonpos.mpr ha).trans ha)] at h
+    rcases h with ⟨rfl, h⟩
+    refine ⟨?_, ?_⟩
+    · simp [h]
+    · intro x
+      simp only [mem_closure_singleton, mem_setOf_eq, and_imp, forall_exists_index]
+      rintro k rfl hk
+      rw [← one_zsmul b, ← mul_zsmul, mul_one, zsmul_le_zsmul_iff h, ← zero_add 1,
+          ← Int.lt_iff_add_one_le]
+      contrapose! hk
+      rw [← Left.nonneg_neg_iff, ← neg_zsmul]
+      exact zsmul_nonneg ha (by simp [hk])
+
+lemma AddSubgroup.mem_closure_singleton_iff_existsUnique_zsmul {a b : G} (ha : a ≠ 0) :
+    (b ∈ closure {a}) ↔ ∃! k : ℤ, k • a = b := by
```

Comment:
This statement is a bit weird. Surely this `ExistsUnique` RHS is hard to prove?

Reply:
@pechersky, was https://github.com/leanprover-community/mathlib4/pull/15659#discussion_r1730314177 meant for here?

Reply:
I guess it's fine to have it like that even if you're only planning on using the `mp` of it, but please then reference the more commonly useful lemma in the docstring of this one